askama_shared/src/generator.rs

```diff
@@ -719,21 +719,58 @@ impl<'a, S: std::hash::BuildHasher> Generator<'a, S> {
         let mut names = Buffer::new(0);
         let mut values = Buffer::new(0);
         for (i, arg) in def.args.iter().enumerate() {
-            if i > 0 {
-                names.write(", ");
-                values.write(", ");
+            let expr = args.get(i).ok_or_else(|| {
+                CompileError::String(format!("macro '{}' takes more than {} arguments", name, i))
+            })?;
+
+            match expr {
+                // If `expr` is already a form of variable then
+                // don't reintroduce a new variable. This is
+                // to avoid moving non-copyable values.
+                Expr::Var(name) => {
+                    let var = self
+                        .locals
+                        .resolve_var(name)
+                        .map(ToString::to_string)
+                        .unwrap_or_else(|| format!("self.{}", name));
```

Comment:
I feel like this is both more compact and easier to understand. What do you think?

```suggestion
                    let var = match self.locals.resolve_var(name) {
                        Ok(s) => s.to_string(),
                        None => format!("self.{}", name)),
                    };
```

Reply:
Needs to be `Some(s)` and you had a `)` to many at the end. But I've applied the change :)